assets/js/archive-status.js

```diff
@@ -0,0 +1,188 @@
+            if (!(this.shouldShowSpinner || this.shouldHideSpinner)) return;
+            let spinner = $(".spinner");
+            if (this.shouldShowSpinner) {
+                spinner.show();
+            } else {
+                spinner.hide();
+            }
+        }
+    }
+
+
+    class CreateArchiveDownloadBar extends DownloadBar {
+        get text() {
+            return "Click here to create a downloadable zip archive for this solution.";
+        }
+
+        get css() {
+            return "list-group-item list-group-item-info";
+        }
+
+        get shouldHideSpinner() {
+            return true;
+        }
+    }
+
+    class PendingArchiveDownloadBar extends DownloadBar {
+        get text() {
+            return "Your archive is being created.";
+        }
+
+        get css() {
+            return "list-group-item list-group-item-info";
+        }
+
+        get shouldShowSpinner() {
+            return true;
+        }
+    }
+
+
+    class AlreadyRunningDownloadBar extends DownloadBar {
+        get text() {
+            return "There is already an archive being created in the background.";
+        }
+
+        get css() {
+            return "list-group-item list-group-item-warning";
+        }
+
+        get shouldShowSpinner() {
+            return true;
+        }
+    }
+
+
+    class DownloadArchiveDownloadBar extends DownloadBar {
+        get text() {
+            return "Click here to download your solution as zip archive.";
+        }
+
+        get css() {
+            return "list-group-item list-group-item-success";
+        }
+
+        get shouldHideSpinner() {
+            return true;
+        }
+
+        prepare() {
+            super.prepare();
+            this.elem.attr("href", SOLUTION_ARCHIVE_DOWNLOAD_URL);
+            this.elem.attr("onClick", "");
+        }
+    }
+
+
+    class DownloadBarFactory {
+        constructor() {
+            this.state = undefined;
+            this.downloadBar = undefined;
+        }
+
+        produce(state) {
+            if (state === this.state) return this.downloadBar;
+            if (state === "available") {
+                this.downloadBar = new DownloadArchiveDownloadBar();
+            } else if (state === "already running") {
+                this.downloadBar = new AlreadyRunningDownloadBar();
+            } else if (state === "initiated") {
+                this.downloadBar = new PendingArchiveDownloadBar();
+            } else {
+                this.downloadBar = new CreateArchiveDownloadBar();
+            }
+            this.state = state;
+            return this.downloadBar;
+        }
+    }
+
+
+    class Endpoint {
+        static get(url, callback) {
+            $.ajax({
+                url: url,
+                type: "get",
+                headers: {"X-CSRFToken": CSRF_TOKEN},
+                datatype: "json",
+                success: callback
+            })
+        }
+    }
+
+
+    createArchive = function () {
```

Comment:
... and this assignment, isn't the best practice in JavaScript to store `createArchive` in the global `window` (or maybe `document`) object, e.g.:

```javascript
window.createArchive = function() { /* ...  */ }
```

(Of course from inside the document.ready boilerplate.)